libyul/optimiser/StackCompressor.cpp

```diff
@@ -253,31 +253,36 @@ bool StackCompressor::run(
 			_optimizeStackAllocation &&
 			evmDialect->evmVersion().canOverchargeGasForCall() &&
 			evmDialect->providesObjectAccess();
-	bool allowMSizeOptimization = !MSizeFinder::containsMSize(_dialect, *_object.code);
+	bool allowMSizeOptimization = !MSizeFinder::containsMSize(_dialect, _object.code->root());
+	Block astRoot = std::get<Block>(ASTCopier{}(_object.code->root()));
 	if (usesOptimizedCodeGenerator)
 	{
-		yul::AsmAnalysisInfo analysisInfo = yul::AsmAnalyzer::analyzeStrictAssertCorrect(_dialect, _object);
-		std::unique_ptr<CFG> cfg = ControlFlowGraphBuilder::build(analysisInfo, _dialect, *_object.code);
+		yul::AsmAnalysisInfo analysisInfo = yul::AsmAnalyzer::analyzeStrictAssertCorrect(_dialect, astRoot, _object.qualifiedDataNames());
+		std::unique_ptr<CFG> cfg = ControlFlowGraphBuilder::build(analysisInfo, _dialect, astRoot);
 		eliminateVariablesOptimizedCodegen(
 			_dialect,
-			*_object.code,
+			astRoot,
 			StackLayoutGenerator::reportStackTooDeep(*cfg),
 			allowMSizeOptimization
 		);
 	}
 	else
+	{
 		for (size_t iterations = 0; iterations < _maxIterations; iterations++)
 		{
-			std::map<YulName, int> stackSurplus = CompilabilityChecker(_dialect, _object, _optimizeStackAllocation).stackDeficit;
+			Object object(_object);
+			object.code = std::make_shared<AST>(std::get<Block>(ASTCopier{}(astRoot)));
+			std::map<YulName, int> stackSurplus = CompilabilityChecker(_dialect, object, _optimizeStackAllocation).stackDeficit;
```

Comment:
Fully agree on immutable analysis info. Then we can be sure it matches.

Reply:
Actually I think it is better to leave it in the `Object` because it directly depends on things in `Object` (particularly data names). But I will set up getter/setter methods for ast so that upon setting a new ast, the analysis info gets nulled